File: packages/libs/SubstrateML/Tests/SubstrateMLTests/CohesionRosterTests.swift
// CohesionRosterTests.swift — pinned vectors shared with cohesion_roster_tests.rs (ADR-026).
import Testing
@testable import SubstrateML

struct CohesionRosterTests {
    let S = CohesionRoster.scale

    @Test("add accumulates exact sums, remove undoes them exactly, replace is remove then add")
    func addRemoveReplace() {
        var r = CohesionRoster()
        r.add(id: "a", digest: "a1", similarities: [])
        r.add(id: "b", digest: "b1", similarities: [S])            // a-b 1.0
        r.add(id: "c", digest: "c1", similarities: [S, S])         // a-c 1.0, b-c 1.0
        r.add(id: "d", digest: "d1", similarities: [0, 0, 0])      // d unlike everyone
        #expect(r.entries.map { [$0.id, "\($0.sum)"] } == [["a", "\(2 * Int64(S))"], ["b", "\(2 * Int64(S))"], ["c", "\(2 * Int64(S))"], ["d", "0"]])
        let before = r
        r.add(id: "e", digest: "e1", similarities: [S / 2, S / 2, S / 2, 0])
        r.remove(id: "e", similarities: [S / 2, S / 2, S / 2, 0])
        #expect(r == before, "remove restores the sums bit for bit")
        var viaReplace = r
        viaReplace.replace(id: "d", digest: "d2", oldSimilarities: [0, 0, 0], newSimilarities: [S, S, S])
        var viaSteps = r
        viaSteps.remove(id: "d", similarities: [0, 0, 0]); viaSteps.add(id: "d", digest: "d2", similarities: [S, S, S])
        #expect(viaReplace == viaSteps)
        #expect(viaReplace.entries.map(\.sum) == [3 * Int64(S), 3 * Int64(S), 3 * Int64(S), 3 * Int64(S)])
    }

    @Test("flags: the unlike member is the outlier at 1.5 sigma and not at 2; small rosters flag nothing")
    func flagsVector() {
        var r = CohesionRoster()
        r.add(id: "a", digest: "a1", similarities: [])
        r.add(id: "b", digest: "b1", similarities: [S])
        r.add(id: "c", digest: "c1", similarities: [S, S])
        r.add(id: "d", digest: "d1", similarities: [0, 0, 0])
        // cohesion a,b,c = 2/3, d = 0; mean 0.5; population sd ≈ 0.2887; z(d) ≈ −1.73
        #expect(r.flags(threshold: 1.5).map(\.anomalous) == [false, false, false, true])
        #expect(r.flags(threshold: 2.0).map(\.anomalous) == [false, false, false, false])
        var small = CohesionRoster()
        small.add(id: "x", digest: "x", similarities: [])
        small.add(id: "y", digest: "y", similarities: [0])
        #expect(small.flags(threshold: 0.1).map(\.anomalous) == [false, false])
    }

    @Test("quantise clamps and rounds to 24-bit fixed point")
    func quantise() {
        #expect(CohesionRoster.quantise(1.0) == S)
        #expect(CohesionRoster.quantise(0.5) == S / 2)
        #expect(CohesionRoster.quantise(-0.2) == 0 && CohesionRoster.quantise(1.7) == S)
        #expect(CohesionRoster.quantise(.nan) == 0)
    }
}

File: packages/libs/SubstrateML/rust/tests/cohesion_roster_tests.rs
// Pinned vectors shared with CohesionRosterTests.swift (ADR-026).
use substrate_ml::cohesion_roster::CohesionRoster;

const S: i32 = CohesionRoster::SCALE;

fn four() -> CohesionRoster {
    let mut r = CohesionRoster::default();
    r.add("a", "a1", &[]);
    r.add("b", "b1", &[S]);
    r.add("c", "c1", &[S, S]);
    r.add("d", "d1", &[0, 0, 0]);
    r
}

#[test]
fn add_remove_replace() {
    let mut r = four();
    let sums: Vec<(String, i64)> = r.entries().iter().map(|e| (e.id.clone(), e.sum)).collect();
    assert_eq!(sums, vec![("a".into(), 2 * S as i64), ("b".into(), 2 * S as i64), ("c".into(), 2 * S as i64), ("d".into(), 0)]);
    let before = r.clone();
    r.add("e", "e1", &[S / 2, S / 2, S / 2, 0]);
    r.remove("e", &[S / 2, S / 2, S / 2, 0]);
    assert_eq!(r, before, "remove restores the sums bit for bit");
    let mut via_replace = r.clone();
    via_replace.replace("d", "d2", &[0, 0, 0], &[S, S, S]);
    let mut via_steps = r.clone();
    via_steps.remove("d", &[0, 0, 0]);
    via_steps.add("d", "d2", &[S, S, S]);
    assert_eq!(via_replace, via_steps);
    assert!(via_replace.entries().iter().all(|e| e.sum == 3 * S as i64));
}

#[test]
fn flags_vector() {
    let r = four();
    let at = |t: f32| -> Vec<bool> { r.flags(t, 3).into_iter().map(|(_, a)| a).collect() };
    assert_eq!(at(1.5), vec![false, false, false, true]);
    assert_eq!(at(2.0), vec![false, false, false, false]);
    let mut small = CohesionRoster::default();
    small.add("x", "x", &[]);
    small.add("y", "y", &[0]);
    assert_eq!(small.flags(0.1, 3).into_iter().map(|(_, a)| a).collect::<Vec<_>>(), vec![false, false]);
}

#[test]
fn quantise_clamps_and_rounds() {
    assert_eq!(CohesionRoster::quantise(1.0), S);
    assert_eq!(CohesionRoster::quantise(0.5), S / 2);
    assert_eq!(CohesionRoster::quantise(-0.2), 0);
    assert_eq!(CohesionRoster::quantise(1.7), S);
    assert_eq!(CohesionRoster::quantise(f32::NAN), 0);
}
